### PR TITLE
Support build cardinal as a plugin

### PR DESCRIPTION
**Pull request by foxspy**
issue: #364 

/kind enhancement

### CODE REVIEW

**Comment by alexanderguzhva** [comment]
Is it a general-purpose plugin system or just a mechanism for optionally attaching Cardinal? If the first one, then how many plugins are we supposed to have? If it is for Cardinal only, then why a plain 
```
#ifdef USE_CARDINAL
#include "cardinal/cardinal.h"
REGISTER_CARDINAL(whatever)
#endif
```
is not sufficient?

**Comment by foxspy** [comment]
> Is it a general-purpose plugin system or just a mechanism for optionally attaching Cardinal? If the first one, then how many plugins are we supposed to have? If it is for Cardinal only, then why a plain
> 
> ```
> #ifdef USE_CARDINAL
> #include "cardinal/cardinal.h"
> REGISTER_CARDINAL(whatever)
> #endif
> ```
> 
> is not sufficient?

We will support not only cardinal. Someone can integrate their own indexes into knowhere without additional development work once their indexes obey the interface rule.

File: src/common/comp/knowhere_config.cc [hunk 1]
```diff
@@ -125,6 +130,79 @@ KnowhereConfig::SetClusteringType(const ClusteringType clustering_type) {
     }
 }
 
+bool
+KnowhereConfig::InitPlugin(std::string path) {
+    static std::once_flag init_plugin_once_;
+    static bool init_result = false;
+
+    auto init = [&]() {
+        auto splitPaths = [&](std::string input, char delimiter) {
```
Comment (alexanderguzhva) [inline src/common/comp/knowhere_config.cc:139]: Please use `std::filesystem` facilities for this

File: src/common/comp/knowhere_config.cc [hunk 2]
```diff
@@ -125,6 +130,79 @@ KnowhereConfig::SetClusteringType(const ClusteringType clustering_type) {
...
+            LOG_KNOWHERE_INFO_ << "Plugin index list is empty";
+        }
+
+        for (auto& lib : libs) {
+            if (!std::filesystem::exists(lib)) {
+                LOG_KNOWHERE_ERROR_ << "Failed to find lib " << lib << " in filesystem";
+                init_result = false;
+                return init_result;
+            }
+            void* plugin = dlopen(lib.c_str(), RTLD_NOW | RTLD_GLOBAL);
```
Comment (alexanderguzhva) [inline src/common/comp/knowhere_config.cc:168]: I could not find the corresponding `dlclose()`

**Review by alexanderguzhva — CHANGES_REQUESTED** [review]
A couple of things to fix. Otherwise, looks good :)

**Pull request closed without merging**